File: tasks/implementations/cpp/tasks/basic_blockchain.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <cstdint>

void run_basic_blockchain(const std::string& size, const std::string& fixturesRoot) {
    constexpr std::uint64_t MOD = 4294967291ULL;
    constexpr std::uint64_t MULT = 16777619ULL;
    std::ifstream file(fixturesRoot + "/generated/blockchain/" + size + ".txt");
    std::string line;
    std::uint64_t prev_hash = 2166136261ULL;
    std::uint64_t total = 0;
    std::uint64_t index = 0;
    while (std::getline(file, line)) {
        if (line.empty()) continue;
        std::stringstream ss(line);
        std::uint64_t hash_value = prev_hash;
        std::uint64_t tx = 0;
        while (ss >> tx) hash_value = (hash_value * MULT + tx + index + 1ULL) % MOD;
        total = (total + hash_value) % MOD;
        prev_hash = hash_value;
        index++;
    }
    std::cout << total << std::endl;
}

File: tasks/implementations/go/tasks/basic_blockchain.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "path/filepath"
    "strconv"
    "strings"
)

func RunBasicBlockchain(size string, fixturesRoot string) {
    const mod uint64 = 4294967291
    const mult uint64 = 16777619
    file, _ := os.Open(filepath.Join(fixturesRoot, "generated", "blockchain", size+".txt"))
    defer file.Close()
    scanner := bufio.NewScanner(file)
    var prevHash uint64 = 2166136261
    var total uint64 = 0
    index := uint64(0)
    for scanner.Scan() {
        line := strings.TrimSpace(scanner.Text())
        if line == "" { continue }
        hashValue := prevHash
        for _, token := range strings.Fields(line) {
            tx, _ := strconv.ParseUint(token, 10, 64)
            hashValue = (hashValue*mult + tx + index + 1) % mod
        }
        total = (total + hashValue) % mod
        prevHash = hashValue
        index++
    }
    fmt.Println(total)
}
